pkg/config/decode.go

```diff
@@ -0,0 +1,114 @@
+package config
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+
+	"github.com/pelletier/go-toml"
+)
+
+// decode parses a TOML tree (and environment variables) according to field tags set in the Config struct.
+//
+// Essentially, it expects `toml` and `env` field tags to be set. Cf. Config for reference.
+func decode(tree *toml.Tree) (conf Config, err error) {
+	// The use of reflection and user-provided data is a very sensitive combination,
+	// searching for "panic" in https://golang.org/pkg/reflect/ shows 82 results.
+	//
+	// Let's be on the safe side and recover from any panic during unmarshalling in order to
+	// return a proper error. Such cases are most likely due to a broken TOML configuration or
+	// invalid env variable value.
+	defer func() {
+		if r := recover(); r != nil {
+			err = r.(error)
+		}
+	}()
+
+	conf = DefaultConfig
+
+	d := &decoder{tree: tree}
+	d.unmarshal(decodeContext{
+		val: reflect.ValueOf(&conf),
+		typ: reflect.TypeOf(&conf),
+	})
+
+	conf.tree = tree
+	return conf, nil
+}
+
+type decoder struct {
+	tree *toml.Tree
+}
+
+type decodeContext struct {
+	val      reflect.Value
+	typ      reflect.Type
+	tomlPath []string
+	envVar   string
+}
+
+func (d *decoder) unmarshal(ctx decodeContext) {
+	// If the current type is a pointer, get the actual type/value behind it.
+	if ctx.typ.Kind() == reflect.Ptr {
+		ctx.typ = ctx.typ.Elem()
+		ctx.val = ctx.val.Elem()
+	}
+
+	switch ctx.typ.Kind() {
+	case reflect.Struct:
+		// The current type is a struct, iterate through its fields.
+		for i := 0; i < ctx.typ.NumField(); i++ {
+			field := ctx.typ.Field(i)
+			// Lookup for the "toml" tag and skip this field if not present.
+			tomlKey, ok := field.Tag.Lookup("toml")
+			if !ok {
+				continue
+			}
+			// Generate a new decodeContext from the struct field and continue unmarshaling.
```

Comment:
Similarly to what you did in `decode`, you can add a blank line before this comment.